File: eduBoostBackend/src/main/java/com/fptu/eduBoostBackend/service/SubscriptionService.java
package com.fptu.eduBoostBackend.service;

import com.fptu.eduBoostBackend.dto.request.InitPaymentRequest;
import com.fptu.eduBoostBackend.dto.response.PaymentTransactionResponse;
import com.fptu.eduBoostBackend.dto.response.RevenueStatsResponse;
import com.fptu.eduBoostBackend.dto.response.SubscriptionPlanResponse;
import com.fptu.eduBoostBackend.dto.response.TeacherSubscriptionResponse;

import java.util.List;

public interface SubscriptionService {

    /** Public — get all active plans */
    List<SubscriptionPlanResponse> getActivePlans();

    /** Teacher — get current active subscription */
    TeacherSubscriptionResponse getMySubscription();

    /** Teacher — initiate payment, trả về thông tin QR để hiển thị */
    PaymentTransactionResponse initiatePayment(InitPaymentRequest request);

    /** Teacher — transaction history */
    List<PaymentTransactionResponse> getMyTransactions();

    /** Teacher — get single transaction by ID */
    PaymentTransactionResponse getTransactionById(Long id);

    /** Admin — all pending transactions */
    List<PaymentTransactionResponse> getPendingTransactions();

    /** Admin — ALL transactions (all statuses) */
    List<PaymentTransactionResponse> getAllTransactions();

    /** Admin — revenue statistics */
    RevenueStatsResponse getRevenueStats();

    /** Admin — confirm a payment and activate subscription */
    TeacherSubscriptionResponse confirmPayment(Long transactionId, String note);

    /** Admin — cancel a pending transaction */
    PaymentTransactionResponse cancelTransaction(Long transactionId, String reason);

    /** VietQR Webhook — auto-confirm payment by orderId */
    boolean confirmPaymentByOrderId(String orderId, long amount);

    /** Scheduled — expire subscriptions past endDate */
    void expireSubscriptions();
}

File: eduBoostFrontend/src/services/subscriptionService.js
import { apiClient } from './api';

const BASE = '/subscriptions';

export const subscriptionService = {
    /** Public — get all active plans for Pricing page */
    getPlans: () => apiClient.get(`${BASE}/plans`).then(r => r.data?.data ?? r.data),

    /** Teacher — get current subscription */
    getMySubscription: () => apiClient.get(`${BASE}/my`).then(r => r.data?.data ?? r.data),

    /** Teacher — initiate payment → get VietQR info */
    initiatePayment: (planId) =>
        apiClient.post(`${BASE}/initiate`, { planId }).then(r => r.data?.data ?? r.data),

    /** Teacher — payment transaction history */
    getMyTransactions: () => apiClient.get(`${BASE}/transactions`).then(r => r.data?.data ?? r.data),

    /** Admin — pending transactions */
    getPendingTransactions: () => apiClient.get(`${BASE}/admin/pending`).then(r => r.data?.data ?? r.data),

    /** Admin — confirm a payment */
    confirmPayment: (id, note = '') =>
        apiClient.post(`${BASE}/admin/confirm/${id}`, { note }).then(r => r.data?.data ?? r.data),

    /** Admin — cancel a pending transaction */
    cancelTransaction: (id, reason = '') =>
        apiClient.post(`${BASE}/admin/cancel/${id}`, { reason }).then(r => r.data?.data ?? r.data),

    /** Poll transaction status by ID (used by payment modal to detect confirmation) */
    getTransactionStatus: (id) =>
        apiClient.get(`${BASE}/transactions/${id}`).then(r => r.data?.data ?? r.data),
};
